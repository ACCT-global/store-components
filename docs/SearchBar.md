# Search Bar

## Description

`Search Bar` is a VTEX Component that shows a search bar with autocomplete options and displays the matching products as well. This component can be imported and used by any VTEX App.

:loudspeaker: **Disclaimer:** Don't fork this project; use, contribute, or open issue with your feature request.

## Table of Contents

- [Usage](#usage)
  - [Props](#props)
  - [CSS Handles](#css-handles)

## Usage

You should follow the usage instruction in the main [README](https://github.com/vtex-apps/store-components/blob/master/README.md#usage).

Then, add `search-bar` block into your app theme, as we do in our [Store Header](https://github.com/vtex-apps/store-header/blob/master/store/blocks.json).

### Props

| Prop name                 | Type                                          | Description                                                                                                                                       | Default value |
| ------------------------- | --------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------- | ------------- |
| `attemptPageTypeSearch`   | `Boolean` | If `true` when clicked on result link of brand, department or category will link to the corresponding brand, department or category page. When `false` will always go to a full text search page.  | `false`        |
| `autocompleteAlignment`   | [`HorizontalAlignment`](#horizontalalignment) | Autocomplete Horizontal alignment.                                                                                                                | `left`        |
| `autocompleteFullWidth`  | `Boolean`                                     | If true, the autocomplete will fill the whole window horizontally.                                                                                | `false`       |
| `autoFocus`               | `Boolean`                                     | Identify if the search input should autofocus or not                                                                                              | -             |
| `blurOnSubmit`            | `Boolean`                                     | Identify if input should blur on submit.                                                                                                          | `false`       |
| `compactMode`             | `Boolean`                                     | Identify when to use the compact version of the component                                                                                         | -             |
| `customSearchPageUrl`     | `string`                                      | Template for a custom url. It can have a substring `${term}` used as placeholder to interpolate the searched term. (e.g. `/search?query=${term}`) | -             |
| `emptyPlaceholder`        | `String!`                                     | Shows a placeholder when the ResultList hasn't results to displayed                                                                               | -             |
| `hasIconLeft`             | `Boolean`                                     | Identify if the search icon is on left or right position                                                                                          | -             |
| ~`iconClasses`~           | `String`                                      | **DEPRECATED** ~Custom classes for the search icon~ Use the CSS handle `searchBarIcon`.                                                           | -             |
| `maxWidth`                | `Number` \| `String`                          | Max width of the search bar                                                                                                                       | -             |
| `minSearchTermLength`     | `Number`                                      | If defined, it will block searches where the term length is lesser than `minSearchTermLength`.                                                    | -             |
| `openAutocompleteOnFocus` | `Boolean`                                     | Identify if autocomplete should be open on input focus or not.                                                                                    | `false`       |
 `placeholder`             | `String!`                                     | Placeholder to be used on the input                                                                                                               | -             |
| `submitOnIconClick`       | `Boolean`                                     | Identify if search icon should submit on click.                                                                                                   | `false`       |

### `HorizontalAlignment`

| Enum name | Enum value |
| --------- | ---------- |
| left      | 'left'     |
| center    | 'center'   |
| right     | 'right'    |

### CSS Handles

Below, we describe the namespace that are defined in the `SearchBar`.

| Class name           | Description                                                           | Component Source                                                                 |
| -------------------- | --------------------------------------------------------------------- | -------------------------------------------------------------------------------- |
| `searchBarContainer` | The main container of `SearchBar`                                     | [SearchBar](/react/components/SearchBar/components/SearchBar.js)                 |
| `compactMode`        | Properties to be applied to the input when `compactMode` prop is true | [AutocompleteInput](/react/components/SearchBar/components/AutocompleteInput.js) |
| `paddingInput`       | The padding of the `SearchBar` input                                  | [AutocompleteInput](/react/components/SearchBar/components/AutocompleteInput.js) |
| `searchBarIcon`      | The class that targets the search bar icon                            | [AutocompleteInput](/react/components/SearchBar/components/AutocompleteInput.js) |
