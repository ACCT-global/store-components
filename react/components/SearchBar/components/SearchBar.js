import React, { Component } from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'
import AutocompleteInput from './AutocompleteInput'
import ResultsLits from './ResultsList'
import DownshiftComponent from 'downshift'
import { NoSSR } from 'render'

export default class SearchBar extends Component {
  render() {
    const {
      placeholder,
      emptyPlaceholder,
      onEnterPress,
      onMakeSearch,
      onInputChange,
      onGoToSearchPage,
      onClearInput,
      shouldSearch,
      inputValue,
      compactMode
    } = this.props

    const fallback = (
      <AutocompleteInput
        placeholder={placeholder}
        onMakeSearch={onMakeSearch}
        onInputChange={onInputChange}
        onGoToSearchPage={onGoToSearchPage}
        inputValue={inputValue}
      />
    )

    const mainClasses = classNames(
      'vtex-searchbar w-100'
    )

    return (
      <div className={mainClasses}>
        <NoSSR onSSR={fallback}>
          <DownshiftComponent>
            {({
              getInputProps,
              selectedItem,
              highlightedIndex,
              isOpen,
              closeMenu,
            }) => (
              <div className="relative-m w-100">
                <AutocompleteInput
                  compactMode={compactMode}
                  onGoToSearchPage={() => {
                    closeMenu()
                    onGoToSearchPage()
                  }}
                  {...getInputProps({
                    placeholder,
                    value: inputValue,
                    onChange: onInputChange,
                    onKeyDown: event => {
                      closeMenu()
                      onEnterPress(event)
                    },
                  })}
                />
                {shouldSearch && isOpen ? (
                  <ResultsLits
                    {...{
                      inputValue,
                      selectedItem,
                      highlightedIndex,
                      emptyPlaceholder,
                      closeMenu,
                      onClearInput,
                    }}
                  />
                ) : null}
              </div>
            )}
          </DownshiftComponent>
        </NoSSR>
      </div>
    )
  }
}

SearchBar.propTypes = {
  /** Empty placeholder to be used on the input */
  emptyPlaceholder: PropTypes.string.isRequired,
  /** Placeholder to be used on the input */
  placeholder: PropTypes.string.isRequired,
  /** Current value of the input */
  inputValue: PropTypes.string.isRequired,
  /** Variable that controls when the search should be made */
  shouldSearch: PropTypes.bool.isRequired,
  /** Function that controls when the search should be executed */
  onMakeSearch: PropTypes.func.isRequired,
  /** Function that controls the action when the enter key is pressed */
  onEnterPress: PropTypes.func.isRequired,
  /** Function to handle input changes */
  onInputChange: PropTypes.func.isRequired,
  /** Function to direct the user to the searchPage */
  onGoToSearchPage: PropTypes.func.isRequired,
  /** Function to clear the input */
  onClearInput: PropTypes.func.isRequired,
  /** Indentify when use the compact version of the component */
  compactMode: PropTypes.bool,
}