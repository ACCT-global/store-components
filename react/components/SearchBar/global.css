.vtex-results__item-image {
  width: 50px;
}

.vtex-results__list {
  list-style-type: none;
  width: calc(100% - 42px);
}

.vtex-results__list a, .vtex-results__list div {
  text-decoration: none;
  color: #727273;
}

.vtex-results__list a:hover {
  background-color: #F2F4F5;
}

.vtex-searchbar input {
  box-shadow: 4px 4px 8px 0 rgba( 0, 0, 0, .2 );
  color: #727273;
}

.vtex-searchbar .vtex-results__list {
  top: 100%;
}

.vtex-search-mobile .vtex-button {
  padding: 0.9rem;
  font-weight: 400;
  font-size: .875rem;
}

.vtex-search-mobile input {
  border: none;
}

@media only screen and (max-width: 49rem) {
  .vtex-searchbar input {
    box-shadow: none;
  }
  .vtex-results__list {
    width: 100%;
  }
}
