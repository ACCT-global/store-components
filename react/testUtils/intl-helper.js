import React from 'react'
import * as reactTestingLibrary from 'react-testing-library'
import { IntlProvider } from 'react-intl'
import defaultStrings from '../../messages/en-US.json'
const Context = React.createContext({})

export const render = (node, options) => {
  const rendered = reactTestingLibrary.render(
    <IntlProvider messages={defaultStrings} locale="en-US">
      {node}
    </IntlProvider>,
    options
  )

  return {
    ...rendered,
    rerender: newUi =>
      render(newUi, {
        container: rendered.container,
        baseElement: rendered.baseElement,
      }),
  }
}

// re-export everything
export default reactTestingLibrary
