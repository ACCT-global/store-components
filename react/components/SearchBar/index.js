import React, { Component } from 'react'
import PropTypes from 'prop-types'
import SearchBar from './components/SearchBar'
import { injectIntl, intlShape } from 'react-intl'

import './global.css'

const SEARCH_DELAY_TIME = 500

/** Canonical search bar that uses the autocomplete endpoint to search for a specific product*/
class SearchBarContainer extends Component {
  timeoutId = null

  state = {
    shouldSearch: false,
    inputValue: '',
  }

  handleClearSearchResults = () => {
    this.setState({ shouldSearch: false })

    this.timeoutId = null
  }

  handleClearInput = () => {
    console.log(this.state.inputValue)
    this.setState({ inputValue: '' })
    this.handleClearSearchResults()
  }

  handleGoToSearchPage = () => {
    const search = this.state.inputValue

    this.setState({ inputValue: '' })
    this.context.navigate({
      page: 'store/search',
      params: { term: search },
      query: 'map=ft',
      fallbackToWindowLocation: false,
    })
  }

  handleEnterPress = event => {
    if (event.key === 'Enter') {
      this.handleGoToSearchPage()
    }
  }

  handleMakeSearch = () => {
    this.handleClearSearchResults()

    this.timeoutId = setTimeout(() => {
      this.setState({ shouldSearch: true })
    }, SEARCH_DELAY_TIME)
  }

  handleInputChange = event => {
    const value = event.target.value

    this.setState({ inputValue: value })

    if (value.length < 2) {
      this.handleClearSearchResults()
    } else this.handleMakeSearch()
  }

  render() {
    const { intl, compactMode } = this.props
    const { shouldSearch, inputValue } = this.state

    const placeholder = intl.formatMessage({
      id: 'search.placeholder',
    })
    const emptyPlaceholder = intl.formatMessage({
      id: 'search.noMatches',
    })

    return (
      <SearchBar
        placeholder={placeholder}
        emptyPlaceholder={emptyPlaceholder}
        shouldSearch={shouldSearch}
        inputValue={inputValue}
        onClearInput={this.handleClearInput}
        onGoToSearchPage={this.handleGoToSearchPage}
        onEnterPress={this.handleEnterPress}
        onMakeSearch={this.handleMakeSearch}
        onInputChange={this.handleInputChange}
        compactMode={compactMode}
      />
    )
  }
}

SearchBarContainer.contextTypes = {
  navigate: PropTypes.func,
}

SearchBarContainer.propTypes = {
  /* Internationalization */
  intl: intlShape.isRequired,
  /** Indentify when use the compact version of the component */
  compactMode: PropTypes.bool,
}



export default injectIntl(SearchBarContainer)
