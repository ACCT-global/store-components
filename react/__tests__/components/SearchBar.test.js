import React from 'react'
import { MockedProvider } from 'react-apollo/test-utils'
import { render } from 'intl-helper'
import { createClientMock } from '../../testUtils/mockProvider'

import SearchBar from '../../SearchBar'

describe('<SearchBar />', () => {
  const mockedResult = {
    data: {
      loading: false,
      autocomplete: {
        itemsReturned: [{ thumb: 'thumb', name: 'name', href: 'href' }],
      },
    },
  }

  const renderComponent = (customProps = {}) => {
    const client = createClientMock(mockedResult)
    return render(
      <MockedProvider client={client}>
        <SearchBar />
      </MockedProvider>
    )
  }

  it('should be able to mount and not break', () => {
    const { asFragment } = renderComponent()
    expect(asFragment()).toBeTruthy()
  })

  it('should match snapshot', () => {
    const { asFragment } = renderComponent()
    expect(asFragment()).toMatchSnapshot()
  })
})
