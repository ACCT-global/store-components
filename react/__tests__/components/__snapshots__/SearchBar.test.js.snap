// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchBar /> should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="w-100 mw7 pv4 searchBarContainer"
  >
    <div
      class="NoSSR-mock"
    >
      <div
        class="relative-m w-100"
      >
        <div
          class="flex"
        >
          <div
            class="w-100"
          >
            <label>
              <input
                aria-autocomplete="list"
                aria-expanded="false"
                autocomplete="off"
                id="downshift-1-input"
                placeholder="Search"
                role="combobox"
                suffix="[object Object]"
                value=""
              />
            </label>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
